add xmp metadata parsing for videos

txl-cp opened the pull request:
SDCP-959

use exiftool to parse xmp tags

to be completed with [client-core feature](https://github.com/superdesk/superdesk-client-core/pull/4882)

taylorli2000 committed:
add xmp metadata write to video on publish

add exiftool dependency to ci

xmp tags as param

fix tag mapping, add test for read from image to xmp

add video type, fix test

fix iptc write with standard iptc tags

fix read from image test

fix tags for image test

fix tagging from photo to video

clean up, add more video tests

reduce fixture filesize

build exiftool for test

stevecanpress committed:
Merge pull request #2 from superdesk/develop

Syncing Superdesk-core develop branch

petrjasek commented on apps/publish/content/common.py:
```diff
@@ -994,15 +1003,28 @@ def _update_picture_metadata(self, updates, original, updated):
 
-                original_metadata = get_metadata_from_item(original, mapping)
-                metadata = get_metadata_from_item(updated, mapping)
+                picture = app.media.get(media_id)
```
this could be a video as well, would call it rather `media`

petrjasek commented on apps/publish/content/common.py:
```diff
@@ -994,15 +1003,28 @@ def _update_picture_metadata(self, updates, original, updated):
 
-                if original_metadata == metadata:
+                file_metadata = (
+                    video_read_metadata(binary) if updated[ITEM_TYPE] == "video" else image_read_metadata(binary)
```
I think it would simplify the usage in general if we would have a generic functions with type param, so this logic could be inside:
```py
file_metadata = read_metadata(binary, updated[item_type])
```
for all those related methods

petrjasek commented on scripts/tests_setup:
```diff
@@ -2,7 +2,16 @@
 # Install python package dependencie
 sudo apt-get -y update
-sudo apt-get -y install libxml2-dev libxmlsec1-dev libxmlsec1-openssl
+sudo apt-get -y install libxml2-dev libxmlsec1-dev libxmlsec1-openssl perl make curl
```
wouldn't installing `libimage-exiftool-perl` be enough?

txl-cp replied:
I tried that initially but I thought it would be better if exiftool versions in core and client-core match. the libimage only installs 12.x

petrjasek replied:
I'm not sure if that's an issue, I suppose most of those tags used are there for long time anyhow and it would simplify the installation in different environments

txl-cp replied:
adjusted to use the libimage

petrjasek commented on superdesk/media/video.py:
```diff
@@ -42,3 +44,161 @@ def get_meta(filestream):
+        logger.exception(e.stderr)
+        return bin
+    finally:
+        remove(path)
```
I guess you could do the whole thing inside that tempfile block,
calling `tmp.seek(0)` after `et.execute` and returning `tmp.read()` after

petrjasek commented on apps/publish/content/common.py:
```diff
@@ -994,15 +1003,28 @@ def _update_picture_metadata(self, updates, original, updated):
+                    video_read_metadata(binary) if updated[ITEM_TYPE] == "video" else image_read_metadata(binary)
+                )
+                metadata = (
+                    get_video_from_photo(get_metadata_from_item(updated, mapping))
```
wouldn't expose that video metadata is converted from photo metadata, that can be done inside the method,
so could be handled inside `get_metadata_from_item` with explicit type param

txl-cp replied:
abstracted to get_metadata_from_item and moved related fns to metadata.py

taylorli2000 committed:
abstract get metadata, read in place for video writes

install exiftool instead of build

specify exiftool version, formatting

black formatting

update exiftool tests version

revert to build exiftool to get latest version

match exiftool version to client-core

match fixture exiftool versions

txl-cp committed:
Merge pull request #9 from txl-cp/feature/video-xmp-metadata

abstract get metadata, read in place for video writes

txl-cp commented on superdesk/media/video.py:
```diff
@@ -42,3 +44,161 @@ def get_meta(filestream):
+
+def map_exiftool_args(video: VideoMetadata) -> list[str]:
+    args = ["-sep", ","]
+    args.extend(f"-{k}={','.join(v) if isinstance(v, list) else v}" for k, v in video.items())
```
don't think this is needed since we're working with standard tags and exiftool knows exacly what tag goes in what namespace

MarkLark86 commented:
@txl-cp Does pyexiftool library provide async capabilities? My concern with Superdesk v3+ is that this will restrict an API instance from processing other requests until the metadata is written into the video file.

petrjasek commented on tests/media/image_metadata_test.py:
```diff
@@ -25,37 +25,41 @@ def image_binary() -> bytes:
-        CopyrightNotice="",
-        CreditLine="The Canadian Press",
-        ProvinceState="PQ",
+        {
```
I like the previous version better, I think copilot got it wrong here, it should work just fine

petrjasek commented on superdesk/media/metadata.py:
```diff
@@ -0,0 +1,54 @@
+    return video_read_metadata(bin) if content_type == "video" else image_read_metadata(bin)
+
+
+def get_metadata_from_item(item: Item, mapping: PhotoMetadataMapping, content_type: CONTENT_TYPE_LITERAL):
```
I wonder if this `PhotoMetadataMapping` should be renamed somehow to reflect it's not only used for photo, would you actually have same config for both photo and video metadata?

txl-cp replied:
combined videometadata and photometadata and moved it over to metadata_mapping

petrjasek commented on superdesk/media/metadata.py:
```diff
@@ -0,0 +1,54 @@
+)
+
+
+def read_metadata(bin: bytes, content_type: CONTENT_TYPE_LITERAL):
```
can you also specify return types in the file?

petrjasek commented on superdesk/media/metadata.py:
```diff
@@ -0,0 +1,54 @@
+    return metadata
+
+
+def get_item_value(item, src: str):
```
would call it `_get_item_value` to make it "private"

txl-cp commented:
> @txl-cp Does pyexiftool library provide async capabilities? My concern with Superdesk v3+ is that this will restrict an API instance from processing other requests until the metadata is written into the video file.

It's not async. I can update the write fn so i/o is non-blocking. is it needed?

MarkLark86 commented:
> > @txl-cp Does pyexiftool library provide async capabilities? My concern with Superdesk v3+ is that this will restrict an API instance from processing other requests until the metadata is written into the video file.
> 
> It's not async. I can update the write fn so i/o is non-blocking. is it needed?

In Superdesk v3.0+ it will be beneficial. Otherwise this task would stop an API/worker instance from processing other API requests while this media processing is going on. So not needed now, but beneficial in v3.0+

taylorli2000 committed:
add types, update tests

petrjasek commented on superdesk/media/metadata_mapping.py:
```diff
@@ -0,0 +1,69 @@
+from typing import Dict, List, Literal, TypedDict
+
+
+class Metadata(TypedDict, total=False):
```
would call it `MediaMetadata` probably, `Metadata` sounds too generic

petrjasek reviewed (commented):
looks good

taylorli2000 committed:
rename metadata to mediametadata

petrjasek commented:
can you check the conflits pls? seems like I don't have permissions

taylorli2000 committed:
Merge remote-tracking branch 'upstream/develop' into feature/video-xmp-metadata
